### ISSUE TITLE
Code Splitting/Async route issue with query string for same route

### ISSUE THREAD

**Issue by just4fun** [mxstbr/react-boilerplate#888]
Hi there,

In the example app, I can search repos for GitHub user. But in my case, I want to share the search result to another person, so query string is needed to indicate specific result.

For example, when I input my GitHub account and press enter in home page, I want the url will be `http://localhost:3000?user=just4fun`. In this way, React Router will notice route changed, which will leads async route for home route be triggered (it has already been triggered when we access home page). And worse still, our sagas will be injected again.

```javascript
path: '/',
name: 'home',
getComponent(nextState, cb) {
  const importModules = Promise.all([
    System.import('containers/HomePage/reducer'),
    System.import('containers/HomePage/sagas'),
    System.import('containers/HomePage'),
  ]);

  const renderRoute = loadModule(cb);

  importModules.then(([reducer, sagas, component]) => {
    injectReducer('home', reducer.default);
    injectSagas(sagas.default);

    renderRoute(component);
  });

  importModules.catch(errorLoading);
},
```

Another side, as for the sagas,

```javascript
export function* githubData() {
  // Fork watcher so we can continue execution
  const watcher = yield fork(getReposWatcher);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}
```

When url updated with query string, `LOCATION_CHANGE` action will be dispatched, it will leads the `watcher` be cancelled.

In conclusion, when I press enter, I did

- update url with query string for enter key handler
- get query string in `mapStateToProps`, then dispatch `loadRepos` action with query string in `componentWillMount` (case: input url directly) and `componentWillReceiveProps` (case: common search)

but nothing happen, b/c

- sagas has been injected again
- watcher in previous sagas has been cancelled

Is there any way to handle this case? Or is there any step I did not in proper way?

**Comment by justingreenberg** [comment]
a simple solution would be to filter location actions by passing function to `take` effect, ie...

```js
export function* githubData() {
  // Fork watcher so we can continue execution
  const watcher = yield fork(getReposWatcher);

  const locationFilter = (action) =>
    action.type === LOCATION_CHANGE && !action.payload.query.user

  // Suspend execution until location changes * without user query *
  yield take(locationFilter);
  yield cancel(watcher);
}
```

now `LOCATION_CHANGE` actions will not cancel watcher:

```js
// eg in homepage component
this.props.changeRoute({ query: { user: 'justin' } })
```

you will still need to decide how to get name into query, whether or not the route drives the logic, etc... hint: you are probably going to want to refactor `getRepos` saga to accept user as param

redux-saga is so flexible that there are probably 50 ways this could be set up :) i hope this at least provides some direction

**Comment by just4fun** [comment]
@justingreenberg - Thank you, I had similar thought with you but I didn't know `take` effect can receive `function`, redux-saga is really very cool.

Now I did same step with you in my sagas, it will avoid my `watcher` to be cancelled.
Then I added more route check in `routes.js`, which will avoid my home stuff be injected again.

```javascript
path: '/',
name: 'home',
getComponent(nextState, cb) {
  if (previousPath === nextState.location.pathname) { return; } // check whether it's same route

  const importModules = Promise.all([
    System.import('containers/HomePage/reducer'),
    System.import('containers/HomePage/sagas'),
    System.import('containers/HomePage'),
  ]);

  const renderRoute = loadModule(cb);

  importModules.then(([reducer, sagas, component]) => {
    injectReducer('home', reducer.default);
    injectSagas(sagas.default);

    renderRoute(component);

    previousPath = nextState.location.pathname; // remember previous route
  });

  importModules.catch(errorLoading);
},
```

**Comment by chaintng** [comment]
@just4fun thanks, your solution work!

**Comment by lsanders** [comment]
I was having a similar problem using [css3 :target selectors](http://www.w3schools.com/cssref/sel_target.asp), say to create a [modal form](http://www.webdesignerdepot.com/2012/10/creating-a-modal-window-with-html5-and-css3/)

Of course the POST saga watcher was getting canceled when navigating "away" from the page (`/someRoute#showForm` to `/someRoute`) 

I managed to figure out how to filter the `LOCATION_CHANGE` take and not cancel the watcher but ended up with an amusing collection of new default sagas and n+1 extra POSTs each time. Now I see how to stop the default saga from being injected over and over - thanks @just4fun!
